File: 2021/python/day13.py
from collections import defaultdict

def drawGrid(G: defaultdict, maxX: int, maxY: int):
    G = defaultdict(lambda: " ", G)
    for y in range(maxY + 1):
        for x in range(maxX + 1):
            print(G[(x, y)], end='')
        print()

points, folds = open("../input/day13.input", encoding="utf-8").read().split("\n\n")
G = {(int(x), int(y)): "#" for x, y in [p.split(",") for p in points.split("\n")]}
folds, A = folds.split("\n"), 0
maxx = max(p[0] for p in G.keys())
maxy = max(p[1] for p in G.keys())

for fold in folds:
    _, _, f = fold.split(" ")
    axis, n = f.split("=")
    if axis == "y":
        maxy //= 2
        G = {(k[0], k[1]) if k[1] < int(n) else (k[0], k[1] - 2 * (k[1] - int(n))): "#"
             for k in G.keys() if k[1] != int(n)}
    else:
        maxx //= 2
        G = {(k[0], k[1]) if k[0] < int(n) else (k[0] - 2 * (k[0] - int(n)), k[1]): "#"
             for k in G.keys() if k[0] != int(n)}
    if A == 0:
        A = len(G.values())

print("Answer A:", A)
print("Answer B:")
drawGrid(G, maxx, maxy)

File: 2021/golang/day13.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func drawSheet(G map[point]string, maxX int, maxY int) {
	for y := 0; y <= maxY; y++ {
		for x := 0; x <= maxX; x++ {
			c := G[point{x: x, y: y}]
			if c == "" {
				c = " "
			}
			fmt.Print(c)
		}
		fmt.Print("\n")
	}
}

func Day13() {
	file, _ := os.Open("../input/day13.input")
	defer file.Close()
	scanner := bufio.NewScanner(file)
	G := make(map[point]string)
	var folds []string
	maxX, maxY, A := 0, 0, 0
	for scanner.Scan() {
		line := scanner.Text()
		if strings.HasPrefix(line, "fold") {
			folds = append(folds, line)
		} else if len(line) > 1 {
			coords := strings.Split(line, ",")
			xcoord, ycoord := coords[0], coords[1]
			x, _ := strconv.Atoi(xcoord)
			y, _ := strconv.Atoi(ycoord)
			G[point{x: x, y: y}] = "#"
			if x > maxX {
				maxX = x
			}
			if y > maxY {
				maxY = y
			}
		}
	}
	for _, fold := range folds {
		parts := strings.Split(fold, "=")
		axis := string(parts[0][-1+len(parts[0])])
		n, _ := strconv.Atoi(parts[1])
		Gnew := make(map[point]string)
		for p := range G {
			if axis == "y" {
				if p.y < n {
					Gnew[p] = "#"
				} else {
					pFlip := point{x: p.x, y: p.y - 2*(p.y-n)}
					Gnew[pFlip] = "#"
				}
			} else {
				if p.x < n {
					Gnew[p] = "#"
				} else {
					pFlip := point{x: p.x - 2*(p.x-n), y: p.y}
					Gnew[pFlip] = "#"
				}
			}
		}
		if axis == "y" {
			maxY /= 2
		} else {
			maxX /= 2
		}
		G = Gnew
		if A == 0 {
			A = len(G)
		}
	}

	print("Answer A:", A)
	print("Answer B:")
	drawSheet(G, maxX, maxY)
}
